src/cargo/ops/tree/graph.rs

```diff
@@ -0,0 +1,562 @@
+//! Code for building the graph used by `cargo tree`.
+
+use super::TreeOptions;
+use crate::core::compiler::{CompileKind, RustcTargetData};
+use crate::core::dependency::DepKind;
+use crate::core::resolver::features::{FeaturesFor, RequestedFeatures, ResolvedFeatures};
+use crate::core::resolver::Resolve;
+use crate::core::{
+    FeatureMap, FeatureValue, InternedString, Package, PackageId, PackageIdSpec, Workspace,
+};
+use crate::util::CargoResult;
+use std::collections::{HashMap, HashSet};
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub enum Node {
+    Package {
+        package_id: PackageId,
+        /// Features that are enabled on this package.
+        features: Vec<InternedString>,
+        kind: CompileKind,
+    },
+    Feature {
+        /// Index of the package node this feature is for.
+        node_index: usize,
+        /// Name of the feature.
+        name: InternedString,
+    },
+}
+
+#[derive(Debug, Copy, Hash, Eq, Clone, PartialEq)]
+pub enum Edge {
+    Dep(DepKind),
+    Feature,
+}
+
+#[derive(Clone)]
+struct Edges(HashMap<Edge, Vec<usize>>);
+
+impl Edges {
+    fn new() -> Edges {
+        Edges(HashMap::new())
+    }
+
+    fn add_edge(&mut self, kind: Edge, index: usize) {
+        let indexes = self.0.entry(kind).or_default();
+        if !indexes.contains(&index) {
+            indexes.push(index)
+        }
+    }
+}
+
+/// A graph of dependencies.
+pub struct Graph<'a> {
+    nodes: Vec<Node>,
+    edges: Vec<Edges>,
```

Comment:
Having only read this far so far, I'm actually not entirely certain how this representation is going to work out. Could this have a comment for how these two lists correspond to each other?

Reply:
(reading a bit more, could this be one list of tuples?)

Reply:
Hm, I tried tuples, but the code ends up being littered with `.0` and `.1` tuple indexes all over, and I think it reduced the readability.  I also kinda tried a struct wrapper, but that also seemed to make it more convoluted.  Dunno, do you feel strongly it needs to change?

Reply:
Nah seems reasonable to me, was just curious! It's documented now so seems fine by me either way.